tests/test_csp_rendering.py

```diff
@@ -0,0 +1,148 @@
+        """Fail if the passed HTML is invalid."""
+        if parser.errors:
+            default_msg = ["Content is invalid HTML:"]
+            lines = content.split(b"\n")
+            for position, error_code, data_vars in parser.errors:
+                default_msg.append("  %s" % E[error_code] % data_vars)
+                default_msg.append("    %r" % lines[position[0] - 1])
+            msg = self._formatMessage(None, "\n".join(default_msg))
+            raise self.failureException(msg)
+
+    @override_settings(
+        MIDDLEWARE=settings.MIDDLEWARE + ["csp.middleware.CSPMiddleware"]
+    )
+    def test_exists(self):
+        """A `nonce` should exist when using the `CSPMiddleware`."""
+        response = self.client.get(path="/regular/basic/")
+        if not isinstance(response, HttpResponse):
+            raise self.failureException(f"{response!r} is not a HttpResponse")
+        self.assertEqual(response.status_code, 200)
+
+        html_root: Element = self.parser.parse(stream=response.content)
+        self._fail_on_invalid_html(content=response.content, parser=self.parser)
+        self.assertContains(response, "djDebug")
+
+        namespaces = get_namespaces(element=html_root)
+        toolbar = list(DebugToolbar._store.values())[0]
+        nonce = str(toolbar.request.csp_nonce)
+        self._fail_if_missing(
+            root=html_root, path=".//link", namespaces=namespaces, nonce=nonce
+        )
+        self._fail_if_missing(
+            root=html_root, path=".//script", namespaces=namespaces, nonce=nonce
+        )
+
+    @override_settings(
+        DEBUG_TOOLBAR_CONFIG={"DISABLE_PANELS": set()},
+        MIDDLEWARE=settings.MIDDLEWARE + ["csp.middleware.CSPMiddleware"],
+    )
+    def test_redirects_exists(self):
+        response = self.client.get("/redirect/")
+        if not isinstance(response, HttpResponse):
+            raise self.failureException(f"{response!r} is not a HttpResponse")
+        self.assertEqual(response.status_code, 200)
+
+        html_root: Element = self.parser.parse(stream=response.content)
+        self._fail_on_invalid_html(content=response.content, parser=self.parser)
+        self.assertContains(response, "djDebug")
+
+        namespaces = get_namespaces(element=html_root)
+        context: ContextList = response.context  # pyright: ignore[reportAttributeAccessIssue]
+        nonce = str(context["toolbar"].request.csp_nonce)
+        self._fail_if_missing(
+            root=html_root, path=".//link", namespaces=namespaces, nonce=nonce
+        )
+        self._fail_if_missing(
+            root=html_root, path=".//script", namespaces=namespaces, nonce=nonce
+        )
+
+    @override_settings(
+        MIDDLEWARE=settings.MIDDLEWARE + ["csp.middleware.CSPMiddleware"]
+    )
+    def test_panel_content_nonce_exists(self):
+        response = self.client.get("/regular/basic/")
+        if not isinstance(response, HttpResponse):
```

Comment:
Can you help me understand the purpose of this check? My understanding is that the test client only returns http responses.

Reply:
It does yes, but the typing is wrong if you don't do this check:

Without, the source lights up in a type checking editor with errors:
![image](https://github.com/user-attachments/assets/a4f98de9-5889-4119-bfce-80fd1db0c2c1)

With:
![image](https://github.com/user-attachments/assets/f719d742-0896-4488-a528-8cb2d3e1d54d)

Basically, type checkers rightly think that walking back on the `self.client.get`, it returns a WSGIRequest and not a HttpResponse. Which is a safe assumption, since the typing is wrong in Django itself: if you go back up to [`django/test/client.py:914`](https://github.com/django/django/blob/7b1a76f899f7a7acb7d70b433cb0064c2184186b/django/test/client.py#L914), you'll see that the type inference rightly assumes that it'll a WSGIRequest, which is not the practical case. 

I didn't delve into it deeper now, but like I said earlier, it is easier to right a wrong here than to go through a very bureucratic process to get this fixed. The syntax itself is valid and ensures that type checking also is right.

Reply:
I'm not a typing expert at all but wouldn't something like `response = typing.cast(self.client.get(...), HttpResponse)` also work, maybe with a comment which says that we're helping the type checker understand the correct type? Otherwise we have code in the testsuite which is (probably) never run. Almost as bad as untested code in the project itself.

Reply:
Ah, so that's your main concern here, the coverage ... I stopped caring about coverage in these minuscule percents a long time ago.

But, be that as it may, I'll change it to using `cast`.

Reply:
No, it's not really about coverage, it's about communicating the intent more clearly. The check as it's written currently communicates that `client.get` might return something which isn't a `HttpResponse`; `typing.cast` communicates that the type checker needs some help here. At least that's my reading of it.

Thanks for your patience. It's not my (our) intention to make the contribution hard. Please assume good intent as well.

Reply:
No offense taken :)

I've already expressed my concerns earlier, not repeating myself meaninglessly.